File: app/schemas/proyecto.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.institucion import InstitucionRead


class ProyectoBase(BaseModel):
    id_proyecto: str
    codigo_proyecto: str | None = None
    nombre: str
    entidad_financiadora: str
    tipo: Literal["investigacion", "extension"] = "investigacion"


class ProyectoCreate(ProyectoBase):
    institucion_ids: list[int] = Field(default_factory=list)


class ProyectoUpdate(BaseModel):
    codigo_proyecto: str | None = None
    nombre: str | None = None
    entidad_financiadora: str | None = None
    tipo: Literal["investigacion", "extension"] | None = None
    institucion_ids: list[int] | None = None


class ProyectoRead(ProyectoBase):
    model_config = ConfigDict(from_attributes=True)

    instituciones: list[InstitucionRead] = Field(default_factory=list)

File: flutter_app/lib/models/proyecto.dart
import 'institucion.dart';

class Proyecto {
  final String idProyecto;
  final String codigoProyecto;
  final String nombre;
  final String entidadFinanciadora;
  final String tipo;
  final List<Institucion> instituciones;

  Proyecto({
    required this.idProyecto,
    required this.codigoProyecto,
    required this.nombre,
    required this.entidadFinanciadora,
    required this.tipo,
    required this.instituciones,
  });

  factory Proyecto.fromJson(Map<String, dynamic> json) {
    final institucionesJson = json['instituciones'];
    return Proyecto(
      idProyecto: json['id_proyecto'] as String,
      codigoProyecto: json['codigo_proyecto'] as String? ?? (json['id_proyecto'] as String),
      nombre: json['nombre'] as String,
      entidadFinanciadora: json['entidad_financiadora'] as String,
      tipo: (json['tipo'] as String?) ?? 'investigacion',
      instituciones: institucionesJson is List
          ? institucionesJson
              .whereType<Map<String, dynamic>>()
              .map(Institucion.fromJson)
              .toList()
          : const [],
    );
  }
}
